README.md
# SSLChecker

## Overview

SSLChecker is a serverless API written in Python and running on Azure Functions. The API is based on Alban Diquet's [SSLyze](https://github.com/nabla-c0d3/sslyze) library. SSLChecker is used to identify obsolete versions of SSL/TLS (e.g., SSL 3.0, and TLS 1.0/1.1) on an endpoint, or perform a full scan to identify all supported versions of SSL/TLS on an endpoint.

## Pre-requisites

__Python__ - Python 3.7 and 3.8 (64-bit). Python 3.9 is not supported yet.

__Development__ - To set up a local development environment, follow the guidance from Microsoft [here](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-first-azure-function-azure-cli?pivots=programming-language-python&tabs=bash%2Cbrowser).

__Deployment__ - As part of the above setup, you will be able to deploy to Azure using the azure-cli. Additionally, Azure DevOps or another CI/CD tool is capable of deploying to Azure.

## Usage

Invoke the function on the command line using curl:

``` curl https://<functionname>.azurewebsite.net/api/{scan:alpha}/{view:alpha}/{target}/{port}```

There are four parts to pass to the URI: scan, view, target, and port.

"scan" is the type of scan: policy or full. Currently, the default policy prohibits using SSL 2.0/3.0, TLS 1.0/1.1 and some TLS 1.2 ciphers, so the policy scan will identify which unsupported ciphers are in use, if any. A full scan will report back all supported ciphers.

Since corporations often use [split-view DNS](https://en.wikipedia.org/wiki/Split-horizon_DNS), "view" in this context is the network viewpoint you want to scan, either internal or external. This is accomplished by specifying a valid DNS server to use for name resolution. The default value for external will use OpenDNS (e.g. 208.67.222.222). The default for internal will be 0.0.0.0 and will result in an error if a scan is attempted and no internal DNS server is specified. Please modify the config.ini file to use an internal DNS server.

"target" should be the DNS domain name or IP that you would like to scan (i.e., github.com or 140.82.113.4).

"port" is optional and if omitted will default to TCP 443.

## A Note on Authentication

Microsoft has extensive documentation on how to secure an HTTP endpoint in Azure Functions [here](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=csharp#secure-an-http-endpoint-in-production). There are two main ways to secure a function: Turn on App Service Authentication/Authorization for the function app, or use Azure API Management (APIM) to authenticate requests. Additionally, Azure functions support API key authorization that you can supply either as a query string variable or in a HTTP header. Microsoft states that API key authorization is not intended as a way to secure an HTTP trigger in production

By default, I have set the authLevel in the function.json file to *anonymous*. Please note, when running functions locally, authorization is disabled regardless of the specified authorization level.

If you plan on running SSLChecker on the internet, please consider one of the above options for authentication.

## Feedback

Send me mail at joe@metlife.com

CHANGELOG.md
# Changelog for SSLChecker

# v1.0.0
- Initial Release

# v1.1.0
- Added the ability to pass a port to SSLScanner

# v1.2.0
- Added some type annotations and more consistent error for the main function

# v1.3.0
- Added the ability to scan by IP

# v2.0.0
- Upgraded [SSLyze](https://github.com/nabla-c0d3/sslyze) to 3.x
- Added several TLS 1.2 ciphers to the "policy" scan type as "weak"
- Added scan type and port to result set

# v2.1.0
- Upgraded dnspython to 2.0.x and fixed deprecated call to dns.resolver.query()
- Upgraded validators to 0.17
- Various pylint and type checking fixes

# v2.2.0
- Upgraded SSLyze to 3.1
- Removed pytest from requirements
- Various pylint and type checking fixes
- Removed TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 and TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 from "strong" TLS 1.2 ciphers

tests/test_SSLChecker.py
""" SSLChecker pytest tests """

import json

import azure.functions as func

import SSLChecker.SSLChecker.main as _main

main = _main.main


def test_policy_external_no_violations():
    """ Test policy scan on an external host with no violations """

    req = func.HttpRequest(
        method='GET',
        body=b'',
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': 'api.metlife.com'}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure there are no violations
    assert results["Results"] == 'No Policy Violations'


def test_full_external():
    """ Test full scan on an external host """
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'full',
                      'view': 'external',
                      'target': 'github.com'}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure there are no violations
    assert results["Results"] != 'No Policy Violations'


def test_policy_external_violations():
    """ Test policy scan on an external host with violations """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': 'espn.com'}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure there are violations
    assert results["Results"] != 'No Policy Violations'


def test_external_dns_name_not_resolved():
    """ Test dns name not resolved """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': 'joegatt.com'}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure the DNS name could not resolve
    assert 'No Answer for joegatt.com using nameserver ' in results["Message"]


def test_external_dns_name_not_exist():
    """ Test NXDOMAIN """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': 'jeogatt.com'}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)
    # Check the output to ensure the DNS name could not resolve
    assert "Domain doesn't exist for jeogatt.com" in results["Message"]


def test_external_sslyze_timeout():
    """ Test sslyze timeout """

    name = 'bbbbbbbbbbbbbbb.com'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': name}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure the DNS name could not resolve
    assert results["Message"] == f'TCP connection to {name}:443 timed-out'


def test_external_missing_target():
    """ Test a request with a missing hostname """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': None}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Ensure error handling is working properly
    assert results['Error Type'] == 'Missing Parameter(s)'
    assert results["Message"] == _main.ERROR_MSG_MISSING_PARAMETERS


def test_bad_dns_view_input():
    """ Test bad dns view input """

    view_name = 'badinput'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': view_name,
                      'target': 'microsoft.com'}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Ensure error handling is working properly
    assert results['Error Type'] == f"Invalid View '{view_name}'"
    assert results["Message"] == _main.ERROR_MSG_INVALID_VIEW


def test_bad_scan_type_input():
    """ Test bad scan type input """

    scan_type = 'pppppp'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': scan_type,
                      'view': 'external',
                      'target': 'microsoft.com'}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Ensure error handling is working properly
    assert results["Error Type"] == f"Invalid scanner type '{scan_type}'"
    assert results["Message"] == _main.ERROR_MSG_INVALID_SCANNER_TYPE


def test_missing_dns_view():
    """ Test not dns view input """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': None,
                      'target': None}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Ensure error handling is working properly
    assert results["Error Type"] == 'Missing Parameter(s)'
    assert results["Message"] == _main.ERROR_MSG_MISSING_PARAMETERS


def test_invalid_dns_name():
    """ Test invalid dns name input """

    dns_name = 'bbbbbbbbb'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': dns_name}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Ensure error handling is working properly
    assert results["Error Type"] == 'Invalid FQDN'
    assert ' is not a valid FQDN' in results["Message"]


def test_missing_policy_view_dns_name():
    """ Test missing scan, view, and target """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': None, # type: ignore
                      'view': None,
                      'target': None}
        )

    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    print(results)
    assert results["Error Type"] == 'Missing Parameter(s)'
    assert results["Message"] == _main.ERROR_MSG_MISSING_PARAMETERS


def test_external_bad_port():
    """ Test bad port input """

    dns_name = 'yahoo.com'
    port = 'a'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': dns_name,
                      'port': port}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure the DNS name could not resolve
    assert results['Error Type'] == f"Invalid Port '{port}'"
    assert results["Message"] == _main.ERROR_MSG_INVALID_PORT


def test_external_port_timeout():
    """ Test timeout connecting to a port """

    dns_name = 'yahoo.com'
    port = '8443'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': dns_name,
                      'port': '8443'}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure the DNS name could not resolve
    assert results['Error Type'] == 'Connection Timeout'
    assert results["Message"] == f'TCP connection to {dns_name}:{port} timed-out'


def test_external_port_not_in_range():
    """ Test port not in valid range """

    port = '123456'
    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': 'espn.com',
                      'port': port}
        )

    # Call the function
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure the DNS name could not resolve
    assert results['Error Type'] == f"Invalid Port '{port}'"
    assert results["Message"] == _main.ERROR_MSG_INVALID_PORT


def test_query_api():
    """ Test query api """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/tls',
        params={'target': 'www.google.com', 'nameserver': '8.8.8.8'}
        )
    resp = main(req)
    assert 'Results' in resp


def test_query_api_by_ip():
    """ Test query api by ip """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/tls',
        params={'target': '140.82.113.4', 'nameserver': '8.8.8.8'}
        )
    resp = main(req)
    assert 'Results' in resp


def test_query_api_error_handling():
    """ Test missing target """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/tls',
        params={'nameserver': '8.8.8.8'}
        )
    resp = main(req)
    results = json.loads(resp)
    assert results['Error Type'] == "Missing required parameter"


def test_policy_external_by_ip_no_violations():
    """ Test policy scan on an external ip with no violations """

    req = func.HttpRequest(
        method='GET',
        body=b"",
        url='/api/',
        route_params={'scan': 'policy',
                      'view': 'external',
                      'target': '216.163.251.205'}
        )
    # Call the function.
    resp = main(req)

    # Convert resp string to dict
    results = json.loads(resp)

    # Check the output to ensure there are violations
    assert results["Results"] == 'No Policy Violations'


SSLChecker/sharedcode/logger.py
"""
the logger module included here is primarily used for local development
it's also useful to ease the creation of file based logger

Unless you want to supply your own formatter, the appropriate default
formatter is used based on the different log level.

Currently, it only means to handle 'debug', 'info', 'warning', 'error'

The file is named as <loggername>_<level>.log, if the loggername ends
with non-alphanumeric characters, it will be sanitized

If no logger name is supplied, it will use the module name as the file
name

Examples
--------
- This create a file myownlogger_info.log, and then log the message
  'whatever' using the default formatter
  ```
    >> import logger
    >> info_logger = logger.get_info_logger('myownlogger')
    >> info_logger.info('whatever')
  ```
- This create a file logger logger_debug.log, and then log the message
  'example' using the default formatter
  ```
    >> import logger
    >> debug_logger = logger.get_debug_logger()
    >> debug_logger.debug('example')
  ```

"""
from typing import Callable, TypeVar, Union
import logging
import string

try:
    from typing import Literal  # only available in python 3.8 and above
except ImportError:
    from typing_extensions import Literal

TLogHandler = TypeVar('TLogHandler', bound=logging.Handler)
TLogger = TypeVar('TLogger', bound=logging.Logger)
TLogLevel = Literal[logging.CRITICAL, logging.ERROR, logging.WARNING,
                  logging.INFO, logging.DEBUG]
TLogFormatter = TypeVar('TLogFormatter', bound=logging.Formatter)

def _setup_logger(logger:TLogger,
                 handler:TLogHandler,
                 formatter:TLogFormatter,
                 log_level:TLogLevel) -> None:
    """
    setup the logger with the handler, log level and formatter
    """
    logger.setLevel(log_level)
    handler.setLevel(log_level)
    handler.setFormatter(formatter)
    logger.addHandler(handler)

_alpha_numerics = string.ascii_letters + string.digits

def sanitize_file_name(name:str) -> str:
    """
    return name that ends with alpha numerics, characters not
    ending in alpha numerics are removed
    """
    last_pos = len(name) - 1
    while name[last_pos] not in _alpha_numerics:
        if last_pos == 0:
            raise ValueError(f'{name} is not valid!')
        last_pos -= 1
    return name[:last_pos + 1]

_log_level_mapping = {'debug': logging.DEBUG,
                     'info': logging.INFO,
                     'warning': logging.WARNING,
                     'error': logging.ERROR}

DEFAULT_DEBUG_LOG_FORMATTER = \
    logging.Formatter('%(asctime)s | %(levelname)s | %(filename)s: %(lineno)5d | %(message)s')
DEFAULT_INFO_LOG_FORMATTER = \
    logging.Formatter('%(asctime)s | %(levelname)s | %(message)s')
DEFAULT_WARNING_LOG_FORMATTER = \
    logging.Formatter('%(asctime)s | WARNING | %(message)s')
DEFAULT_ERROR_LOG_FORMATTER = \
    logging.Formatter('%(asctime)s | ERROR | %(message)s')

FILE_HANDLERS = {}
def level_logger_factory(
        log_level:Literal['debug', 'info', 'warning', 'error'] = 'info',
        log_formatter:TLogFormatter = DEFAULT_INFO_LOG_FORMATTER
    ) -> Callable[[str, TLogFormatter], TLogger]:
    """
    a file based logger factory for the appropriate log level
    """
    def _get_logger(logger_name:str = __name__,
                    log_formatter:TLogFormatter = log_formatter
                    ) -> TLogger:
        """
        return the appropriate logger
        """

        log_file_name = f'{sanitize_file_name(logger_name)}_{log_level}.log'

        log_handler = FILE_HANDLERS.get(log_file_name)
        use_new_handler = False
        if log_handler is None:
            log_handler = logging.FileHandler(log_file_name)
            FILE_HANDLERS[log_file_name] = log_handler
            use_new_handler = True

        logger = logging.getLogger(logger_name)
        if use_new_handler:
            _setup_logger(logger, log_handler, log_formatter,
                         _log_level_mapping[log_level])
        return logger

    return _get_logger

get_info_logger = level_logger_factory()
get_debug_logger = level_logger_factory('debug', DEFAULT_DEBUG_LOG_FORMATTER)
get_warning_logger = level_logger_factory('warning', DEFAULT_WARNING_LOG_FORMATTER)
get_error_logger = level_logger_factory('error', DEFAULT_ERROR_LOG_FORMATTER)


SSLChecker/sharedcode/scanner.py
from hashlib import md5
from typing import Dict, Any

from sslyze import (
    ServerNetworkLocationViaDirectConnection,
    ServerConnectivityTester,
    errors,
    ScanCommand,
    Scanner,
    ServerScanRequest,
)

from . import results
from .errors import ConnectionError

# Policy prohibits the use of SSL 2.0/3.0, TLS 1.0/1.1 and
# some TLS 1.2 cipher suites
CIPHER_SUITES = {
    "policy": [
        ScanCommand.SSL_2_0_CIPHER_SUITES,
        ScanCommand.SSL_3_0_CIPHER_SUITES,
        ScanCommand.TLS_1_0_CIPHER_SUITES,
        ScanCommand.TLS_1_1_CIPHER_SUITES,
        ScanCommand.TLS_1_2_CIPHER_SUITES,
    ],
    "full": [
        ScanCommand.SSL_2_0_CIPHER_SUITES,
        ScanCommand.SSL_3_0_CIPHER_SUITES,
        ScanCommand.TLS_1_0_CIPHER_SUITES,
        ScanCommand.TLS_1_1_CIPHER_SUITES,
        ScanCommand.TLS_1_2_CIPHER_SUITES,
        ScanCommand.TLS_1_3_CIPHER_SUITES,
    ],
}

# Currently, only The following TLS 1.2 ciphers are considered "strong"
OK_TLS12_CIPHERS = {
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
    "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
}


ERROR_MSG_CONNECTION_TIMEOUT = "TCP connection to {}:{} timed-out".format
ERROR_MSG_UNKNOWN_CONNECTION = (
    "TCP connection to {}:{} encountered unknown error".format
)


def scan(target, ip, port, view, suite) -> Dict[str, Any]:
    """ Five inputs: web site name, ip, port
    split-dns view, and cipher suite """

    server_location = ServerNetworkLocationViaDirectConnection(target, port, ip)

    # This line checks to see if the host is online
    try:
        server_info = ServerConnectivityTester().perform(server_location)
    except errors.ConnectionToServerTimedOut:
        raise ConnectionError(
            "Connection Timeout", ERROR_MSG_CONNECTION_TIMEOUT(target, port)
        )
    except errors.ConnectionToServerFailed:
        raise ConnectionError(
            "Unknown Connection Error", ERROR_MSG_UNKNOWN_CONNECTION(target, port)
        )

    # Create a new results dictionary
    scan_output = results.new_result_set()

    # I hash the combination of hostname and ip for tracking
    key = md5((target + ip).encode("utf-8")).hexdigest()
    results.set_result(scan_output, "MD5", key)
    results.set_result(scan_output, "Target", f"{target}:{port}")
    results.set_result(scan_output, "IP", f"{ip}:{port}")
    results.set_result(scan_output, "Scan", suite)
    results.set_result(scan_output, "View", view)

    scanner = Scanner()
    # Ignore type error on get(key) as it defaults to None
    # https://docs.python.org/3/library/stdtypes.html#dict.get
    # We supply the values in the dict
    server_scan_req = ServerScanRequest(
        server_info=server_info, scan_commands=CIPHER_SUITES.get(suite)  # type: ignore
    )
    scanner.queue_scan(server_scan_req)

    for result in scanner.get_results():
        for cipher_suite in CIPHER_SUITES.get(suite):
            scan_result = result.scan_commands_results[cipher_suite]

            for accepted_cipher_suite in scan_result.accepted_cipher_suites:
                if suite == "policy" and scan_result.tls_version_used.name == "TLS_1_2":
                    if (
                        accepted_cipher_suite.cipher_suite.name
                        not in OK_TLS12_CIPHERS
                    ):
                        results.set_ciphers(
                            scan_output,
                            {
                                "Version": f"{scan_result.tls_version_used.name}",
                                "Cipher": f"{accepted_cipher_suite.cipher_suite.name}",
                            },
                        )
                else:
                    results.set_ciphers(
                        scan_output,
                        {
                            "Version": f"{scan_result.tls_version_used.name}",
                            "Cipher": f"{accepted_cipher_suite.cipher_suite.name}",
                        },
                    )

    if len(scan_output["Results"]) == 0:
        results.set_result(scan_output, "Results", "No Policy Violations")

    return scan_output


SSLChecker/sharedcode/shared_dns.py
""" Shared module for dns operations """

import pathlib
from configparser import ConfigParser, ParsingError
from typing import Tuple
from urllib.parse import urlparse

from dns import resolver
from validators import domain
from .errors import InvalidConfig, InvalidFQDN, UnknownError, DNSError

# Alias str type to better reflect the intented type and value
Fqdn = str
IpAddress = str


def get_dns_options() -> Tuple[str, str]:
    """
    Parse config.ini file and get internal and external dns servers
    """

    # https://github.com/Azure/azure-functions-python-worker/issues/340
    config_file = pathlib.Path(__file__).parent / "config.ini"
    parser = ConfigParser()

    try:
        parser.read_string(config_file.read_text())
        external_dns = parser.get("dns_view", "external")
        internal_dns = parser.get("dns_view", "internal")
    except ParsingError as err:
        raise InvalidConfig(
            f"Invalid Configuration File {config_file.resolve()}", err
        )

    return external_dns, internal_dns


def _init_resolver(dnsserver: IpAddress, timeout: int, lifetime: int) -> resolver.Resolver:
    """
    initialize a resolver
    """
    custom_resolver = resolver.Resolver(configure=False)
    custom_resolver.timeout = timeout
    custom_resolver.lifetime = lifetime
    custom_resolver.nameservers = [dnsserver]
    return custom_resolver


def resolve_dns(
        dnsserver: IpAddress, dnsname: Fqdn,
        timeout: int = 3, lifetime: int = 3
        ) -> IpAddress:
    """ Resolve dns name """
    _iplist = []  # results

    res = _init_resolver(dnsserver, timeout, lifetime)

    try:
        answers = res.resolve(dnsname, search=False)  # explicit query for A record
        for answer in answers.rrset:
            _iplist.append(answer.address)
        return _iplist[0]  # Return the first IP of the DNS Answer

    except resolver.NoAnswer:
        raise DNSError(
            "DNS No Answer", f"No Answer for {dnsname} using nameserver {dnsserver}"
        )
    except resolver.NXDOMAIN:
        raise DNSError(
            "DNS Non-Existing Domain",
            (f"Domain doesn't exist for {dnsname} using nameserver {dnsserver}"),
        )
    except resolver.Timeout:
        raise DNSError(
            "DNS operation timed out",
            (
                f"Operation not completed for {dnsname} within "
                f"{lifetime} using nameserver {dnsserver}"
            ),
        )

    # you don't know if the DNS server is truely
    # offline or network condition preventing the DNS request to get to the
    # server, it's best to just use another custom error for this condition
    except Exception:
        raise UnknownError(
            "DNS Unknown Error",
            (
                f"Error encountered while resolving {dnsname} "
                f"using nameserver {dnsserver}"
            ),
        )


def parse_name(name: str) -> Fqdn:
    """ Parse a DNS name to ensure it does not contain http(s) """
    parsed_name = urlparse(name)

    # The below parses out http(s) from a name
    dns_name_candidate = parsed_name.netloc
    if dns_name_candidate == "":
        dns_name_candidate = parsed_name.path

    # The below ensures a valid domain was supplied
    if not domain(dns_name_candidate):
        raise InvalidFQDN("Invalid FQDN", f"{name} is not a valid FQDN")

    return dns_name_candidate


SSLChecker/SSLChecker/main.py
import ipaddress
import json
import logging
from typing import Tuple, Mapping
from time import process_time
from urllib.parse import urlparse

import azure.functions as func

from ..sharedcode import shared_dns
from ..sharedcode import scanner
from ..sharedcode import results
from ..sharedcode.errors import (InvalidRequest, DNSError, InvalidConfig,
                                 ConnectionError, InvalidFQDN)

EXTERNAL_DNS, INTERNAL_DNS = shared_dns.get_dns_options()

DNSVIEW = {"external": EXTERNAL_DNS,
           "internal": INTERNAL_DNS}

# Valid scan types
VALID_SCAN_TYPES = ['policy', 'full']

ERROR_MSG_MISSING_PARAMETERS = \
    ("Please pass three parameters in the URI: "
     "valid scan type: policy or full, valid DNS view: internal or external, "
     "and a valid DNS domain name or IP. For example: "
     "https://<functionname>.azurewebsite.net/api/full/external/github.com")

ERROR_MSG_INVALID_SCANNER_TYPE = \
    "Please pass a valid scan type: 'policy' or 'full'"

ERROR_MSG_INVALID_VIEW = \
    "Please pass a valid DNS view: internal or external"

ERROR_MSG_MISSING_DNS_SERVER = \
    "Please specify a valid DNS server in config.ini"

ERROR_MSG_INVALID_DNS_NAME = \
    "Not a valid formatted DNS name"

ERROR_MSG_INVALID_PORT = \
    "Please pass a valid port in range 1-65535"


def verify_port(port: str) -> int:
    """
    raises InvalidRequest
    """
    if port.isnumeric() is False:
        raise InvalidRequest(f"Invalid Port '{port}'", ERROR_MSG_INVALID_PORT)
    _port = int(port)
    if _port > 65535 or _port == 0:
        raise InvalidRequest(f"Invalid Port '{port}'", ERROR_MSG_INVALID_PORT)
    return _port


def verify_scan_type(scan_type: str) -> str:
    """
    verify scan type is valid
    """
    scan_type = scan_type.lower()
    if scan_type not in VALID_SCAN_TYPES:
        raise InvalidRequest(f"Invalid scanner type '{scan_type}'",
                             ERROR_MSG_INVALID_SCANNER_TYPE)
    return scan_type


def pre_scan_check(req: func.HttpRequest) -> Tuple[str, str, str, int, str]:
    """
    return scan_type, view, target, port, and ip as a tuple if the request is
    valid.

    raises
    ------
    InvalidRequest
    InvalidConfig
    DNSError
    """
    scan_type = req.route_params.get('scan')
    view = req.route_params.get('view')
    target = req.route_params.get('target')
    # Default to TCP 443 if no port is passed
    port = req.route_params.get('port', '443')
    port = verify_port(port)

    """ Check to ensure ALL parameters were passed in the URI.
    If you mark the route parameters in function.json as mandatory,
     the Azure Function worker supplies a 404 if you do not supply all
     three routes in the URI. I made routes optional, this way we
     can handle errors gracefully """
    if scan_type is None or view is None or target is None:
        raise InvalidRequest("Missing Parameter(s)",
                             ERROR_MSG_MISSING_PARAMETERS)

    scan_type = verify_scan_type(scan_type)
    # Check to ensure a valid DNS view was passed
    view = view.lower()
    if view not in DNSVIEW:
        raise InvalidRequest(f"Invalid View '{view}'", ERROR_MSG_INVALID_VIEW)

    # this maybe best handled as part of loading the app instead of checking it
    # here
    if DNSVIEW.get(view) == '0.0.0.0':
        raise InvalidConfig('Missing DNS Server in config',
                            ERROR_MSG_MISSING_DNS_SERVER)

    # See if a valid IP was passed, else check if it was a valid DNS name
    ip = ""

    try:
        if ipaddress.IPv4Address(target):
            ip = target
    except:
        # Parse the target parameter to ensure it is a valid FQDN
        # and does not contain http(s)
        target = shared_dns.parse_name(target)

        """ Try to resolve the DNS name to an IP to ensure it exists.
        We use the IP in the scan so that we can record which one we tested
        which can be useful. """

        # Ignore type error on get(key) as it defaults to None
        # https://docs.python.org/3/library/stdtypes.html#dict.get
        # I check that the key exists
        ip = shared_dns.resolve_dns(DNSVIEW.get(view), target)  # type: ignore

    return scan_type, view, target, port, ip


def path_params_scanner(req: func.HttpRequest) -> str:
    """
    perform actual scan for path based parameters
    """
    try:
        scan_type, view, target, port, ip = pre_scan_check(req)
        # Run the scan
        return json.dumps(scanner.scan(target, ip, port, view, scan_type))
    except (InvalidRequest, InvalidConfig, DNSError, ConnectionError,
            InvalidFQDN) as err:
        return json.dumps(results.set_error(err.args[0], err.args[1]))
    except Exception as err:
        return json.dumps(results.set_error("Unexpected Error", str(err)))


VALID_QUERY_PARAMS = ('target', 'nameserver', 'port', 'scan_type')

ERROR_MSG_QUERY_EXAMPLE = (
    "Example: https://<functionname>.azurewebsite.net/api/ssl?"
    "target=www.yahoo.com&port=8443")

ERROR_MSG_INVALID_QUERY_PARAMS = ("Valid params are: "
                                  f"{', '.join(VALID_QUERY_PARAMS)}")
ERROR_MSG_INVALID_QUERY_URL = ("Valid URL path must be 'ssl' or 'tls'. "
                               f"{ERROR_MSG_QUERY_EXAMPLE}")
ERROR_MSG_INVALID_QUERY_MISSING_PARAMS = ("'target' parameters is required. "
                                          f"{ERROR_MSG_QUERY_EXAMPLE}")


def query_scanner_precheck(url: str,
                           params: Mapping[str, str]
                           ) -> Tuple[str, str, int, str, str, str]:
    """
    check to ensure the url path as well as the query parameters are valid

    returns
    -------
    scan_type, target, port, ip, nameserver, view

    raises
    ------
    InvalidRequest
    """

    if url.lower() not in VALID_QUERY_API_URL:
        raise InvalidRequest(f"Invalid URL Path '{url}'",
                             ERROR_MSG_INVALID_QUERY_URL)
    if 'target' not in params:
        raise InvalidRequest(f'Missing required parameter',
                             ERROR_MSG_INVALID_QUERY_MISSING_PARAMS)

    target = params['target']
    scan_type = 'full'
    port = 443
    nameserver = None

    for param in params:
        if param not in VALID_QUERY_PARAMS:
            raise InvalidRequest(f"Invalid Parameter supplied '{param}'",
                                 ERROR_MSG_INVALID_QUERY_PARAMS)
        if param == 'scan_type':
            scan_type = verify_scan_type(params[param])
        elif param == 'nameserver':
            nameserver = params[param]
        elif param == 'port':
            port = verify_port(params[param])

    view = 'custom'
    if nameserver is None:
        if EXTERNAL_DNS:
            view = 'external'
            nameserver = EXTERNAL_DNS
        else:
            view = 'internal'
            nameserver = INTERNAL_DNS

    # See if a valid IP was passed, else check if it was a valid DNS name
    ip = ""

    try:
        if ipaddress.IPv4Address(target):
            ip = target
    except:
        target = shared_dns.parse_name(target)
        ip = shared_dns.resolve_dns(nameserver, target)

    return scan_type, target, port, ip, nameserver, view


VALID_QUERY_API_URL = ('ssl', 'tls')


def query_params_scanner(url: str, params: Mapping[str, str]) -> str:
    """
    new function behavior to handle query based scanner, it would default
    to external DNS view before using the Internal
    """
    try:
        scan_type, target, port, ip, nameserver, view = \
            query_scanner_precheck(url, params)
        return json.dumps(scanner.scan(target, ip, port, view, scan_type))
    except (InvalidRequest, InvalidConfig, DNSError, ConnectionError,
            InvalidFQDN) as err:
        return json.dumps(results.set_error(err.args[0], err.args[1]))
    except Exception as err:
        return json.dumps(results.set_error("Unexpected Error", str(err)))


def main(req: func.HttpRequest) -> str:
    logging.info('Python HTTP trigger function processed a request '
                 f'for url: {req.url}.')
    starttime = process_time()
    url_parsed = urlparse(req.url)
    url_path = [v for v in url_parsed.path.split('/')[2:] if v]

    if not url_path or len(url_path) > 1 or url_path[0] in VALID_SCAN_TYPES:
        resp = path_params_scanner(req)
    else:
        resp = query_params_scanner(url_path[0], req.params)

    logging.info(f'Processing time for URL {req.url} took'
                 f' {process_time() - starttime}')

    return resp
